tests/test_wasm_intrinsics_simd.c

```diff
@@ -0,0 +1,1476 @@
+}
+v128_t TESTFN i16x8_sub(v128_t x, v128_t y) {
+  return wasm_i16x8_sub(x, y);
+}
+v128_t TESTFN i16x8_sub_saturate_s(v128_t x, v128_t y) {
+  return wasm_i16x8_sub_saturate(x,y);
+}
+v128_t TESTFN i16x8_sub_saturate_u(v128_t x, v128_t y) {
+  return wasm_u16x8_sub_saturate(x, y);
+}
+v128_t TESTFN i16x8_mul(v128_t x, v128_t y) {
+  return wasm_i16x8_mul(x, y);
+}
+v128_t TESTFN i32x4_neg(v128_t vec) {
+  return wasm_i32x4_neg(vec);
+}
+int32_t TESTFN i32x4_any_true(v128_t vec) {
+  return wasm_i32x4_any_true(vec);
+}
+int32_t TESTFN i32x4_all_true(v128_t vec) {
+  return wasm_i32x4_all_true(vec);
+}
+v128_t TESTFN i32x4_shl(v128_t vec, int32_t shift) {
+  return wasm_i32x4_shl(vec, shift);
+}
+v128_t TESTFN i32x4_shr_s(v128_t vec, int32_t shift) {
+  return wasm_i32x4_shr(vec, shift);
+}
+v128_t TESTFN i32x4_shr_u(v128_t vec, int32_t shift) {
+  return wasm_u32x4_shr(vec, shift);
+}
+v128_t TESTFN i32x4_add(v128_t x, v128_t y) {
+  return wasm_i32x4_add(x, y);
+}
+v128_t TESTFN i32x4_sub(v128_t x, v128_t y) {
+  return wasm_i32x4_sub(x, y);
+}
+v128_t TESTFN i32x4_mul(v128_t x, v128_t y) {
+  return wasm_i32x4_mul(x, y);
+}
+
+#ifdef __wasm_unimplemented_simd128__
+
+
+v128_t TESTFN i64x2_neg(v128_t vec) {
+  return wasm_i64x2_neg(vec);
+}
+
+#ifdef __wasm_unimplemented_simd128__
+
+bool TESTFN i64x2_any_true(v128_t vec) {
+  return wasm_i64x2_any_true(vec);
+}
+bool TESTFN i64x2_all_true(v128_t vec) {
+  return wasm_i64x2_all_true(vec);
+}
+
+#endif // __wasm_unimplemented_simd128__
+
+v128_t TESTFN i64x2_shl(v128_t vec, int32_t shift) {
+  return wasm_i64x2_shl(vec, shift);
+}
+
+v128_t TESTFN i64x2_shr_s(v128_t vec, int32_t shift) {
+  return wasm_i64x2_shr(vec, shift);
+}
+v128_t TESTFN i64x2_shr_u(v128_t vec, int32_t shift) {
+  return wasm_u64x2_shr(vec, shift);
+}
+v128_t TESTFN i64x2_add(v128_t x, v128_t y) {
+  return wasm_i64x2_add(x, y);
+}
+v128_t TESTFN i64x2_sub(v128_t x, v128_t y) {
+  return wasm_i64x2_sub(x, y);
+}
+
+#endif // __wasm_unimplemented_simd128__
+
+v128_t TESTFN f32x4_abs(v128_t vec) {
+  return wasm_f32x4_abs(vec);
+}
+v128_t TESTFN f32x4_neg(v128_t vec) {
+  return wasm_f32x4_neg(vec);
+}
+
+#ifdef __wasm_unimplemented_simd128__
+
+v128_t TESTFN f32x4_sqrt(v128_t vec) {
+  return wasm_f32x4_sqrt(vec);
+}
+
+#endif // __wasm_unimplemented_simd128__
+
+v128_t TESTFN f32x4_add(v128_t x, v128_t y) {
+  return wasm_f32x4_add(x,y);
+}
+v128_t TESTFN f32x4_sub(v128_t x, v128_t y) {
+  return wasm_f32x4_sub(x, y);
+}
+v128_t TESTFN f32x4_mul(v128_t x, v128_t y) {
+  return wasm_f32x4_mul(x, y);
+}
+v128_t TESTFN f32x4_div(v128_t x, v128_t y) {
+  return wasm_f32x4_div(x, y);
+}
+v128_t TESTFN f32x4_min(v128_t x, v128_t y) {
+  return wasm_f32x4_min(x, y);
+}
+v128_t TESTFN f32x4_max(v128_t x, v128_t y) {
+  return wasm_f32x4_max(x, y);
+}
+
+#ifdef __wasm_unimplemented_simd128__
+
+v128_t TESTFN f64x2_abs(v128_t vec) {
+  return wasm_f64x2_abs(vec);
+}
+v128_t TESTFN f64x2_neg(v128_t vec) {
+  return -vec;
```

Comment:
Should we use wasm_f64x2_neg? Current way of direct arithmetic operations cause failures in our local experiment.
Similar for f64x2_add, f64x2_sub, etc.

Reply:
@tlively wrapping `v128_t` in a struct without methods (e.g. without `operator-`) would have prevented this bug. I expect these types of bugs to be very common, and only show up when you try to compile with a different implementation.

Reply:
Point taken. This particular bug occurred because these tests were originally copied from the existing emscripten tests that use the vector extensions directly. Real users will not be converting non-portable code in this manner so I do not expect practical issues with this.

Reply:
I agree with @jing-bao, test should use the intrinsic, that is what matters from developer's point of view, while compilers can do different things _underneath_.

Reply:
Yes, this was a copy-paste error. All of these test functions should be using the intrinsics.